presto-main/src/main/java/com/facebook/presto/operator/project/DictionaryAwarePageFilter.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.project;
+
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.LazyBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+
+import java.util.Optional;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+public class DictionaryAwarePageFilter
```

Comment:
There is a lot less overlap then you would think. They have very different return types and methods.

Also, in a general sense, I really disk like abstract classes (and inheritance hierarchies) because it is difficult for me to read the code, since I have to switch back and forth between multiple files.